segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,295 +139,572 @@
     }
 
     /**
      * Commits given transaction.
      *
      * @param txn       transaction to commit.
      * @param lazyWrite true if data can be written lazily.
-     * @throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
+     * @return A CompletableFuture that, when completed, will indicate that the operation completed.
+     * If the operation failed, it will contain the cause of the failure. Notable exceptions:
+     * {@link StorageMetadataException} if transaction can not be committed.
      */
     @Override
-    public void commit(MetadataTransaction txn, boolean lazyWrite) throws StorageMetadataException {
-        commit(txn, lazyWrite, false);
+    public CompletableFuture<Void> commit(MetadataTransaction txn, boolean lazyWrite) {
+        return commit(txn, lazyWrite, false);
     }
 
     /**
      * Commits given transaction.
      *
      * @param txn transaction to commit.
-     * @throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
+     * @return A CompletableFuture that, when completed, will indicate that the operation completed.
+     * If the operation failed, it will contain the cause of the failure. Notable exceptions:
+     * {@link StorageMetadataException} if transaction can not be committed.
      */
     @Override
-    public void commit(MetadataTransaction txn) throws StorageMetadataException {
-        commit(txn, false, false);
+    public CompletableFuture<Void> commit(MetadataTransaction txn) {
+        return commit(txn, false, false);
     }
 
     /**
      * Commits given transaction.
      *
      * @param txn       transaction to commit.
      * @param lazyWrite true if data can be written lazily.
-     * @throws StorageMetadataException StorageMetadataVersionMismatchException if transaction can not be commited.
+     * @return A CompletableFuture that, when completed, will indicate that the operation completed.
+     * If the operation failed, it will contain the cause of the failure. Notable exceptions:
+     * {@link StorageMetadataException} if transaction can not be committed.
      */
     @Override
-    public void commit(MetadataTransaction txn, boolean lazyWrite, boolean skipStoreCheck) throws StorageMetadataException {
+    public CompletableFuture<Void> commit(MetadataTransaction txn, boolean lazyWrite, boolean skipStoreCheck) {
         Preconditions.checkArgument(null != txn);
-        if (fenced.get()) {
-            throw new StorageMetadataWritesFencedOutException("Transaction writer is fenced off.");
-        }
-
-        Map<String, TransactionData> txnData = txn.getData();
+        val txnData = txn.getData();
+
+        val modifiedKeys = new ArrayList<String>();
+        val modifiedValues = new ArrayList<TransactionData>();
+        val t = new Timer();
+        val retValue = CompletableFuture.runAsync(() -> {
```

Comment:
You can rewrite this as 
```
if(fenced.get()){
retval = Futures.failedFuture(new StorageMetadataWritesFencedOutException(...)) ;
}else{
 .. put the `thenComposeAsync` body in here
}
```

Reply:
This may require a bit more fiddling around with that `whenComplete`, but at least it should look a bit better.